### ISSUE TITLE
Windows long path support

### ISSUE THREAD

**Issue by HertzDevil** [crystal-lang/crystal#13420]
By default, [filenames on Windows cannot be longer than 256 characters, excluding the drive name](https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation). There are two ways to overcome this since Windows 10 version 1607:

* Replace `path` with `"\\\\?\\#{File.expand_path(path)}"` at the Win32 API boundary;
* Set the `Computer\HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\FileSystem\LongPathsEnabled` registry key to 1 and link a manifest containing the `longPathAware` element into every program.

MSVC can only embed one manifest into an executable, so if the user provides their own manifest, they have to merge their own manifest with the one generated by Crystal, or Crystal's CLI has to support manifests specifically. There is also no guarantee that the registry key can be set by a user. Therefore, the first option is probably more suitable for the standard library, as all we need to do is trying the extended name if the original name fails:

```crystal
module Crystal::System::Dir
  def self.create(path : String, mode : Int32) : Nil
    if LibC._wmkdir(System.to_wstr(path)) == -1
      if LibC._wmkdir(System.to_wstr("\\\\?\\#{File.expand_path(path)}")) == -1
        raise ::File::Error.from_errno("Unable to create directory", file: path)
      end
    end
  end
end
```

Although not mentioned in the linked article, it seems C runtime functions also support long paths in this manner. It might be better to switch to their Win32 equivalents whenever possible though (`CreateDirectoryW` in this example).

**Comment by HertzDevil** [comment]
`path.size > 256` is not a necessary condition here, because at this point `path` could still be a relative path and the current working directory could have been near the `MAX_PATH` limit.

**Comment by stakach** [comment]
reading the docs, the `"\\\\?\\#{File.expand_path(path)}"` at the boundary should work in older versions of windows too
the registry key and manifest are only required if you want long paths using the common Win32 file and directory functions
versus the Unicode versions to permit an extended-length path

**Comment by HertzDevil** [comment]
Note that [contrary to the docs](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-setcurrentdirectory), `SetCurrentDirectoryW` does not work for long paths, and when you are inside a long directory in the command prompt, you can't start any processes at all.

**Comment by stakach** [comment]
the command prompt has issues with long paths. The cmd.exe process does not support long paths, even with the manifest file - need to use powershell

I think the only workaround is to use paths with the `\\?\` prefix

**Comment by HertzDevil** [comment]
.NET does its own thing, but the same error happens as soon as it tries to do non-.NET things:

![image](https://github.com/crystal-lang/crystal/assets/1361918/1394ff63-0ed0-45c5-9e42-bf470953040e)

What I'm saying is long paths aren't supported here even with `\\?\`.

**Comment by stakach** [comment]
I think that's unrelated as when you try to start a new process in a directory with a path that's too long, the new process will inherit the terminals current directory, which would cause problems if the path is longer than 260 characters.

**Comment by mominshaikhdevs** [comment]
`CreateProcessW` does not support long paths at all (both `\\?\` and manifest files). 

Those APIs need to support long paths in future Windows versions.

**Comment by stakach** [comment]
Fairly certain that the application can be at a long path location, that is lpApplicationName can be a long path.
It's only the lpCurrentDirectory param that does not support long paths.

**Comment by TwoLeaves** [comment]
Is this issue the cause of the error I'm seeing when trying to compile a Crystal program on Windows? And is there a workaround? The error is as below (experienced when compiling natively and cross-compiling from Linux):

```
In /usr/lib/crystal/crystal/system/win32/process.cr:210:48

 210 | make_env_block(env, clear_env), chdir.try { |str| System.to_wstr(str) },
                                             ^--
Error: argument 'lpCurrentDirectory' of 'LibC#CreateProcessW' must be Pointer(UInt16), not (Pointer(UInt16) | Nil)
```

**Comment by straight-shoota** [comment]
@TwoLeaves That's an entirely unrelated error. I have created a new issue for that: #13767

**Comment by straight-shoota** [comment]
I stumbled across this thread on the topic: https://mastodon.gamedev.place/@AshleyGullen/111109299141510319

Related HN discussion: https://news.ycombinator.com/item?id=37632232

**Comment by stakach** [comment]
yeah it looks like you can do this to set the code page programmatically:

```
#include <Windows.h>

int APIENTRY WinMain(HINSTANCE hInstance,
                     HINSTANCE hPrevInstance,
                     LPSTR    lpCmdLine,
                     int       nCmdShow)
{
    // Set the process code page to UTF-8
    SetThreadLocale(LOCALE_CUSTOM_UNSPECIFIED);
    SetThreadUILanguage(LANG_NEUTRAL);
    SetConsoleOutputCP(CP_UTF8);
    SetConsoleCP(CP_UTF8);

    // Now you can use the "A" functions with UTF-8 strings.
    MessageBoxA(NULL, "Hello, 世界!", "UTF-8 Message", MB_OK);

    return 0;
}
```

**Comment by mominshaikhdevs** [comment]
* https://github.com/microsoft/WindowsAppSDK/pull/5830#issuecomment-3300274024.
